### PR TITLE
introduce WC_Round_Tax class

### LINKED ISSUE #24619: Cumulated rounding error on net payment
All prices in our shop are entered in gross value. When selling into non EU countries only the net price is charged. WooCommerce starts to calculate the net price by dividing the gross price by the VAT-factor. 

Lets say we have a product for 9,95 Euro in the Shop (gross value.) WC divides by 1,19 (19% VAT) and the result is 8.36134453... and rounds it to 8.36 (net value). The problem now is, when the customer orders 5 of this. 5 x 8.36 Euro results in 41.80 Euro but in the basket and invoice WC states that it is 41.81 Euro. That is because in this case WC calculates the complete gross value first 5 x 9,95 Euro (49.75) and divides this value by 1,19. The result is 41,806722... which will correctly by rounded to 41,81 Euro. You see the problem. Single product price 8.36 Euro. Buying 5 of it should be 41.80 Euro not 41.81 Euro which is not correct.

**Screenshots**
Bestellung_bearbeiten_‹_Marisco_Naturkosmetik_—_WordPress

**Isolating the problem (mark completed items with an [x]):**
- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [ ] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>
`
### WordPress Environment ###

WordPress address (URL): https://www.marisco-naturkosmetik.de
Site address (URL): https://www.marisco-naturkosmetik.de
WC Version: 3.6.5
Log Directory Writable: ✔
WP Version: 5.2.3
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: –
WP Cron: ✔
Language: de_DE_formal
External object cache: –

### Server Environment ###

Server Info: nginx/1.17.1
PHP Version: 7.2.21-1+ubuntu16.04.1+deb.sury.org+1
PHP Post Max Size: 2 GB
PHP Time Limit: 30
PHP Max Input Vars: 10000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.2.26-MariaDB-10.2.26+maria~xenial-log
Max Upload Size: 2 GB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.6.5
WC Database Prefix: wp_oihqaskqjd_
MaxMind GeoIP Database: ✔
Datenbank-Gesamtgröße: 570.95MB
Datenbank-Datengröße: 325.86MB
Datenbank-Indexgröße: 245.09MB
wp_oihqaskqjd_woocommerce_sessions: Daten: 22.06MB + Index: 0.33MB
wp_oihqaskqjd_woocommerce_api_keys: Daten: 0.02MB + Index: 0.03MB
wp_oihqaskqjd_woocommerce_attribute_taxonomies: Daten: 0.02MB + Index: 0.02MB
wp_oihqaskqjd_woocommerce_downloadable_product_permissions: Daten: 0.02MB + Index: 0.06MB
wp_oihqaskqjd_woocommerce_order_items: Daten: 0.23MB + Index: 0.08MB
wp_oihqaskqjd_woocommerce_order_itemmeta: Daten: 3.52MB + Index: 3.03MB
wp_oihqaskqjd_woocommerce_tax_rates: Daten: 0.02MB + Index: 0.06MB
wp_oihqaskqjd_woocommerce_tax_rate_locations: Daten: 0.02MB + Index: 0.03MB
wp_oihqaskqjd_woocommerce_shipping_zones: Daten: 0.02MB + Index: 0.00MB
wp_oihqaskqjd_woocommerce_shipping_zone_locations: Daten: 0.02MB + Index: 0.03MB
wp_oihqaskqjd_woocommerce_shipping_zone_methods: Daten: 0.02MB + Index: 0.00MB
wp_oihqaskqjd_woocommerce_payment_tokens: Daten: 0.02MB + Index: 0.02MB
wp_oihqaskqjd_woocommerce_payment_tokenmeta: Daten: 0.02MB + Index: 0.03MB
wp_oihqaskqjd_woocommerce_log: Daten: 0.02MB + Index: 0.02MB
wp_oihqaskqjd_commentmeta: Daten: 0.08MB + Index: 0.09MB
wp_oihqaskqjd_comments: Daten: 1.52MB + Index: 0.55MB
wp_oihqaskqjd_feedmanager_channel: Daten: 0.02MB + Index: 0.03MB
wp_oihqaskqjd_feedmanager_country: Daten: 0.02MB + Index: 0.03MB
wp_oihqaskqjd_feedmanager_errors: Daten: 0.02MB + Index: 0.00MB
wp_oihqaskqjd_feedmanager_feed_status: Daten: 0.02MB + Index: 0.08MB
wp_oihqaskqjd_feedmanager_field_categories: Daten: 0.02MB + Index: 0.00MB
wp_oihqaskqjd_feedmanager_product_feed: Daten: 0.02MB + Index: 0.02MB
wp_oihqaskqjd_feedmanager_product_feedmeta: Daten: 0.02MB + Index: 0.03MB
wp_oihqaskqjd_feedmanager_source: Daten: 0.02MB + Index: 0.08MB
wp_oihqaskqjd_icl_content_status: Daten: 0.02MB + Index: 0.02MB
wp_oihqaskqjd_icl_core_status: Daten: 0.02MB + Index: 0.02MB
wp_oihqaskqjd_icl_flags: Daten: 0.02MB + Index: 0.02MB
wp_oihqaskqjd_icl_languages: Daten: 0.02MB + Index: 0.03MB
wp_oihqaskqjd_icl_languages_translations: Daten: 0.19MB + Index: 0.08MB
wp_oihqaskqjd_icl_locale_map: Daten: 0.02MB + Index: 0.00MB
wp_oihqaskqjd_icl_message_status: Daten: 0.02MB + Index: 0.03MB
wp_oihqaskqjd_icl_mo_files_domains: Daten: 0.05MB + Index: 0.02MB
wp_oihqaskqjd_icl_node: Daten: 0.02MB + Index: 0.00MB
wp_oihqaskqjd_icl_reminders: Daten: 0.02MB + Index: 0.00MB
wp_oihqaskqjd_icl_strings: Daten: 5.52MB + Index: 4.97MB
wp_oihqaskqjd_icl_string_packages: Daten: 0.02MB + Index: 0.00MB
wp_oihqaskqjd_icl_string_pages: Daten: 221.75MB + Index: 211.34MB
wp_oihqaskqjd_icl_string_positions: Daten: 0.48MB + Index: 0.13MB
wp_oihqaskqjd_icl_string_status: Daten: 0.02MB + Index: 0.02MB
wp_oihqaskqjd_icl_string_translations: Daten: 3.52MB + Index: 1.52MB
wp_oihqaskqjd_icl_string_urls: Daten: 0.45MB + Index: 0.38MB
wp_oihqaskqjd_icl_translate: Daten: 8.52MB + Index: 1.52MB
wp_oihqaskqjd_icl_translate_job: Daten: 0.16MB + Index: 0.06MB
wp_oihqaskqjd_icl_translations: Daten: 0.48MB + Index: 0.94MB
wp_oihqaskqjd_icl_translation_batches: Daten: 0.02MB + Index: 0.00MB
wp_oihqaskqjd_icl_translation_status: Daten: 13.52MB + Index: 0.09MB
wp_oihqaskqjd_links: Daten: 0.02MB + Index: 0.02MB
wp_oihqaskqjd_options: Daten: 2.52MB + Index: 0.31MB
wp_oihqaskqjd_postmeta: Daten: 26.55MB + Index: 11.03MB
wp_oihqaskqjd_posts: Daten: 5.52MB + Index: 1.06MB
wp_oihqaskqjd_redirection_404: Daten: 0.02MB + Index: 0.06MB
wp_oihqaskqjd_redirection_groups: Daten: 0.02MB + Index: 0.03MB
wp_oihqaskqjd_redirection_items: Daten: 5.52MB + Index: 4.98MB
wp_oihqaskqjd_redirection_logs: Daten: 1.52MB + Index: 0.66MB
wp_oihqaskqjd_smush_dir_images: Daten: 0.02MB + Index: 0.03MB
wp_oihqaskqjd_termmeta: Daten: 0.06MB + Index: 0.06MB
wp_oihqaskqjd_terms: Daten: 0.06MB + Index: 0.03MB
wp_oihqaskqjd_term_relationships: Daten: 0.48MB + Index: 0.27MB
wp_oihqaskqjd_term_taxonomy: Daten: 0.05MB + Index: 0.03MB
wp_oihqaskqjd_usermeta: Daten: 0.31MB + Index: 0.25MB
wp_oihqaskqjd_users: Daten: 0.02MB + Index: 0.05MB
wp_oihqaskqjd_wc_download_log: Daten: 0.02MB + Index: 0.03MB
wp_oihqaskqjd_wc_product_meta_lookup: Daten: 0.16MB + Index: 0.31MB
wp_oihqaskqjd_wc_webhooks: Daten: 0.02MB + Index: 0.02MB
wp_oihqaskqjd_yoast_seo_links: Daten: 0.02MB + Index: 0.02MB
wp_oihqaskqjd_yoast_seo_meta: Daten: 0.30MB + Index: 0.00MB

### Post Type Counts ###

attachment: 4364
invoice: 1198
nav_menu_item: 109
oembed_cache: 1
page: 30
post: 2
product: 1726
product_variation: 2
revision: 228
shop_order: 630
shop_order_refund: 68
wpps-shortcodes: 8

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (29) ###

ManageWP - Worker: von GoDaddy – 4.9.1
WPML Multilingual CMS: von OnTheGoSystems – 4.2.7.1
All-in-One WP Migration: von ServMask – 7.1
Angell EYE Updater: von Angell EYE – 2.0.6
Antispam Bee: von pluginkollektiv – 2.9.1
BackUpWordPress: von XIBO Ltd – 3.10
Classic Editor: von WordPress-Mitwirkende – 1.5
GDPR Cookie Consent: von WebToffee – 1.7.7
Duplicate Post: von Enrico Battocchi – 3.2.2
Google Tag Manager for Wordpress: von Thomas Geiger – 1.9.2 – Nicht getestest mit der aktiven WooCommerce-Version
WPBakery Page Builder: von Michael M - WPBakery.com – 6.0.3
Umleitungen: von John Godley – 4.3.1
woo-inquire-us-and-disable-add-to-cart-button: von themelocation – 1.3
Vendidero Helper: von Vendidero – 1.2.1
WooCommerce Blöcke: von Automattic – 2.2.0
PayPal PLUS für WooCommerce: von Inpsyde GmbH – 2.0.4
Germanized für WooCommerce Pro: von Vendidero – 1.11.1
Germanized für WooCommerce: von Vendidero – 2.3.2
WooCommerce Multilingual: von OnTheGoSystems – 4.6.5
Sofort Banking for WooCommerce: von Sven Wagener – 1.4.3
WooCommerce: von Automattic – 3.6.5 – 3.7.0 ist verfügbar
Yoast SEO: von Team Yoast – 11.6 – Nicht getestest mit der aktiven WooCommerce-Version
WP Product Feed Manager: von Michel Jongbloed – 1.14.0
Protected Shops: von conpark GmbH – 2.0
Protected Shops für WooCommerce: von conpark GmbH – 2.0 – Nicht getestest mit der aktiven WooCommerce-Version
Smush: von WPMU DEV – 3.2.1
WPML Media: von OnTheGoSystems – 2.5.2
WPML String Translation: von OnTheGoSystems – 2.10.5.1
WPML Translation Management: von OnTheGoSystems – 2.8.6.2

### Must Use Plugins (1) ###

ManageWP - Worker Loader: von ManageWP –

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: EUR (€)
Currency Position: right_space
Thousand Separator: .
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop-Basis: #25 - /
Warenkorb: #6 - /warenkorb/
Kasse: #7 - /kasse/
Mein Konto: #8 - /mein-konto/
Allgemeine Geschäftsbedingungen: #44 - /agb/

### Theme ###

Name: MNV1
Version: 1 License: GNU General Public
Author URL: https://www.udo-telaar.de
Child Theme: ✔
Parent Theme Name: VendiPro
Parent Theme Version: 1.7.7
Parent Theme Author URL: http://wordpress.org/
WooCommerce Support: ✔

### Templates ###

Overrides: mnv1/woocommerce/archive-product.php
vendipro/woocommerce/cart/mini-cart.php
vendipro/woocommerce/checkout/form-checkout.php
vendipro/woocommerce/content-widget-product.php
vendipro/woocommerce/content-widget-reviews.php
mnv1/woocommerce/emails/customer-note.php
mnv1/woocommerce/emails/customer-processing-order.php
mnv1/woocommerce/emails/plain/customer-processing-order.php
vendipro/woocommerce/global/wrapper-end.php
vendipro/woocommerce/global/wrapper-start.php
vendipro/product-searchform.php
vendipro/woocommerce/single-product/review.php

### Action Scheduler ###

Complete: 0
Oldest: –
Newest: –

Pending: 0
Oldest: –
Newest: –

Canceled: 0
Oldest: –
Newest: –

In-progress: 0
Oldest: –
Newest: –

Failed: 0
Oldest: –
Newest: –

`

</details>

### PR DESCRIPTION
**Pull request by rrennick**
### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #24619 .

### How to test the changes in this Pull Request:

1.
2.
3.

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### CODE REVIEW

**Comment by rrennick** [comment]
@vedanshujain Can you have a look at this class before I continue the PR?

File: includes/class-wc-round-tax.php [hunk 1]
```diff
@@ -0,0 +1,183 @@
+<?php
```
Comment (vedanshujain) [inline includes/class-wc-round-tax.php:1]: Should we move it t the new PSR4 `src` directory?

File: includes/class-wc-round-tax.php [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+	 * @var bool
+	 */
+	protected $inclusive = false;
+
+	/**
+	 * Tax rate.
+	 *
+	 * @var float
+	 */
+	protected $tax_rate = 0.0;
```
Comment (vedanshujain) [inline includes/class-wc-round-tax.php:34]: will probably need to support multiple tax rates with compounding

**Review by vedanshujain — COMMENTED** [review]
I like the approach 👍 . Are you also planning to add a new class for calculating totals and combining different tax and subtotal values as well?

File: includes/class-wc-round-tax.php (hunk 1 above)
Comment (rrennick) [reply includes/class-wc-round-tax.php:1]: Yes, do we need to wait for #26731 to land?

File: includes/class-wc-round-tax.php (hunk 2 above)
Comment (rrennick) [reply includes/class-wc-round-tax.php:34]: > Are you also planning to add a new class for calculating totals and combining different tax and subtotal values as well?

That seems like the best approach.

File: includes/class-wc-round-tax.php (hunk 1 above)
Comment (claudiosanches) [reply includes/class-wc-round-tax.php:1]: I think we could move now.

**Comment by rrennick** [comment]
@vedanshujain I'm going to close this PR but leave the branch as the work to date might be useful in the full solution.

**Pull request closed without merging**